src/memory.rs

```diff
@@ -533,12 +554,13 @@ mod tests {
 	#[test]
 	fn zero_copy() {
 		let mem = MemoryInstance::alloc(Pages(1), None).unwrap();
+		mem.set(100, &[0]).expect("memory set should not fail");
 		mem.with_direct_access_mut(|buf| {
-			assert_eq!(buf.len(), 65536);
+			assert_eq!(buf.len(), 101);
```

Comment:
Oh, actually this can be surprising for the user. So the user must account that the amount of allocated memory may or may not be equal to the length of the buffer. 

Furthermore, the user might want to differentiate between out-of-bounds access (this should trap) and just untouched memory. The complication here is that we can't really tell the difference between uninitialized memory and zeroed memory because all memory is implicitly zeroed. So it is possible that some host function (aka "externalities func") can be called on (fully or partially) untouched  region. 

On top of that, in the worst case the user might not be able to write anything to the `buf`, because the region that is inteded to write just have never been touched and basically the user should fallback to non-direct access, that IMO complicates matters a lot.

Reply:
They can resize the buffer as they need. The percieved allocated size is returned in `current_size` so it is easy enough to to differentiate between out-of-bounds access.

The user should not be using such low level API in the first place. There's a tradeoff between providing a low level API and making efficient abstractions.

Reply:
>  There's a tradeoff between providing a low level API and making efficient abstractions.

100% indeed users should already use this function with caution since it might lead to panics. But I wonder if there is a better approach to this, e.g. returning wrapper struct that handles such details? 

Anyway, I wouldn't block the PR from merge and I think a plain comment that cautions the user about this special thing would do.

Reply:
I've added a comment and changed `with_direct_access_mut` so that the closure can control allocation.